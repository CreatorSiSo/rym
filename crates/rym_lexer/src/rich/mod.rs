use rym_unescape::unquote;
use Mapped::{Multi, Single};

mod test;
mod token;
pub use token::*;

use crate::primitive;

type Span = std::ops::Range<usize>;

#[derive(Clone, Debug)]
pub struct Lexer<'a> {
	/// Absolute offset within the source of the current character.
	pos: usize,
	/// Source text to tokenize.
	src: &'a str,
	/// Cursor for getting lexer tokens.
	primitive_lexer: primitive::Lexer<'a>,
}

impl<'a> Lexer<'a> {
	pub fn new(src: &'a str) -> Self {
		Self { pos: 0, src, primitive_lexer: primitive::Lexer::new(src) }
	}

	fn bump(&mut self) -> Option<(primitive::TokenKind, Span)> {
		self.primitive_lexer.next_token().map(|token| {
			let start_pos = self.pos;
			self.pos += token.len as usize;
			(token.kind, start_pos..self.pos)
		})
	}

	fn peek(&self) -> Option<(primitive::TokenKind, Span)> {
		self.primitive_lexer.clone().next_token().map(|token| {
			let end_pos = self.pos + token.len as usize;
			(token.kind, self.pos..end_pos)
		})
	}

	fn src_from_span(&self, span: &Span) -> &'a str {
		&self.src[span.start..span.end]
	}

	fn map_next(
		&mut self,
		span: &mut Span,
		map_fn: impl Fn(Option<primitive::TokenKind>) -> Mapped,
	) -> Token {
		let mut end = span.end;
		match if let Some((kind, end_span)) = self.peek() {
			end = end_span.end;
			map_fn(Some(kind))
		} else {
			map_fn(None)
		} {
			Mapped::Multi(kind) => {
				*span = span.start..end;
				self.bump();
				kind
			}
			Mapped::Single(kind) => kind,
		}
	}
}

enum Mapped {
	Single(Token),
	Multi(Token),
}

impl Iterator for Lexer<'_> {
	type Item = (Token, Span);

	fn next(&mut self) -> Option<Self::Item> {
		while let Some((primitive_kind, mut span)) = self.bump() {
			let kind = match primitive_kind {
				primitive::TokenKind::Whitespace => {
					if self.src_from_span(&span).contains('\n') {
						Token::Newline
					} else {
						continue;
					}
				}
				primitive::TokenKind::LineComment => continue,
				primitive::TokenKind::BlockComment { terminated } => {
					if terminated {
						continue;
					}
					// TODO: Special reporting for nested block comments
					// self.handler.emit(
					// 	Diagnostic::new_spanned(Level::Error, "Unterminated block comment", span)
					// 		.sub_diagnostic(
					// 			Level::Note,
					// 			None,
					// 			"Missing trailing `*/` to terminate the block comment",
					// 		),
					// );
					continue;
				}

				// Punctuation
				primitive::TokenKind::Semi => Token::Semi,
				primitive::TokenKind::Colon => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Colon) => Multi(Token::ColonColon),
					_ => Single(Token::Colon),
				}),
				primitive::TokenKind::Dot => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Dot) => Multi(Token::DotDot),
					_ => Single(Token::Dot),
				}),
				primitive::TokenKind::Comma => Token::Comma,

				// Operator like
				primitive::TokenKind::Or => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Or) => Multi(Token::OrOr),
					_ => Single(Token::Or),
				}),
				primitive::TokenKind::And => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::And) => Multi(Token::AndAnd),
					_ => Single(Token::And),
				}),
				primitive::TokenKind::Plus => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Eq) => Multi(Token::PlusEq),
					_ => Single(Token::Plus),
				}),
				primitive::TokenKind::Minus => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Eq) => Multi(Token::MinusEq),
					Some(primitive::TokenKind::GreaterThan) => Multi(Token::ThinArrow),
					_ => Single(Token::Minus),
				}),
				primitive::TokenKind::Star => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Eq) => Multi(Token::StarEq),
					_ => Single(Token::Star),
				}),
				primitive::TokenKind::Slash => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Eq) => Multi(Token::SlashEq),
					_ => Single(Token::Slash),
				}),
				primitive::TokenKind::Percent => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Eq) => Multi(Token::PercentEq),
					_ => Single(Token::Percent),
				}),
				primitive::TokenKind::Eq => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Eq) => Multi(Token::EqEq),
					Some(primitive::TokenKind::GreaterThan) => Multi(Token::FatArrow),
					_ => Single(Token::Eq),
				}),
				primitive::TokenKind::Bang => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Eq) => Multi(Token::BangEq),
					_ => Single(Token::Bang),
				}),
				primitive::TokenKind::LessThan => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Eq) => Multi(Token::LessThanEq),
					_ => Single(Token::LessThan),
				}),
				primitive::TokenKind::GreaterThan => self.map_next(&mut span, |kind| match kind {
					Some(primitive::TokenKind::Eq) => Multi(Token::GreaterThanEq),
					_ => Single(Token::GreaterThan),
				}),

				// Delimiter
				primitive::TokenKind::OpenParen => Token::OpenParen,
				primitive::TokenKind::CloseParen => Token::CloseParen,
				primitive::TokenKind::OpenBrace => Token::OpenBrace,
				primitive::TokenKind::CloseBrace => Token::CloseBrace,
				primitive::TokenKind::OpenBracket => Token::OpenBracket,
				primitive::TokenKind::CloseBracket => Token::CloseBracket,

				// Indentifier or Keyword
				primitive::TokenKind::Ident => {
					let name = self.src_from_span(&span);
					match KEYWORDS_MAP.0.iter().position(|kw| *kw == name) {
						Some(pos) => KEYWORDS_MAP.1[pos].clone(),
						None => Token::Ident(name.to_string()),
					}
				}

				primitive::TokenKind::Integer => Token::Int(
					match self
						.src_from_span(&span)
						.chars()
						.filter(|c| c != &'_')
						.collect::<String>()
						.parse::<i64>()
					{
						Ok(int) => int,
						Err(err) => unreachable!(
							"Internal Error: Could not parse <i64> from {0}: `{err}`",
							self.src_from_span(&span)
						),
					},
				),
				primitive::TokenKind::Float => {
					let Some((l, r)) = self.src_from_span(&span).split_once('.') else {
							unreachable!("Internal Error: Float literal does not contain a '.'")
						};
					let l_val: u32 = l.parse().expect(&format!(
						"Internal Error: Left hand side of float literal has invalid value `{l}`"
					));
					let r_val: u32 = r.parse().unwrap_or(0);
					Token::Float(l_val, r_val)
				}
				primitive::TokenKind::Char { terminated } => {
					if !terminated {
						todo!();
						// TODO
						// self.handler.emit(
						// 	Diagnostic::new_spanned(Level::Error, "Unterminated character literal", span)
						// 		.sub_diagnostic(
						// 			Level::Note,
						// 			None,
						// 			"Missing trailing `'` to terminate the character literal",
						// 		),
						// );
						// continue;
					}
					let string = match unquote(self.src_from_span(&span)) {
						Ok(string) => string,
						Err(err) => todo!("Unquote error: {err}"),
					};
					Token::Char(match string.parse::<char>() {
						Ok(char) => char,
						Err(_err) => {
							todo!();
							// TODO
							// self.handler.emit(Diagnostic::new_spanned(
							// 	Level::Error,
							// 	format!("Could not parse <char>: {err}"),
							// 	span,
							// ));
							// continue;
						}
					})
				}
				primitive::TokenKind::String { terminated } => {
					if !terminated {
						// TODO
						// self.handler.emit(
						// 	Diagnostic::new_spanned(Level::Error, "Unterminated string literal", span)
						// 		.sub_diagnostic(
						// 			Level::Note,
						// 			None,
						// 			"Missing trailing `\"` to terminate the string literal",
						// 		),
						// );
						continue;
					}
					let string = match unquote(self.src_from_span(&span)) {
						Ok(string) => string,
						Err(_err) => {
							// TODO
							// self.handler.emit(Diagnostic::new_spanned(Level::Error, err.to_string(), span));
							continue;
						}
					};
					Token::String(string)
				}

				primitive::TokenKind::Unkown => {
					// TODO
					// self.handler.emit(Diagnostic::new_spanned(Level::Error, "Invalid character", span));
					continue;
				}
				primitive::TokenKind::At
				| primitive::TokenKind::Caret
				| primitive::TokenKind::Dollar
				| primitive::TokenKind::Pound
				| primitive::TokenKind::Tilde
				| primitive::TokenKind::Question => {
					// TODO
					// self.handler.emit(Diagnostic::new_spanned(Level::Error, "Reserved character", span));
					continue;
				}
			};

			return Some((kind, span));
		}
		None
	}
}
