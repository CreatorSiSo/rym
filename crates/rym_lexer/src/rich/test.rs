#![cfg(test)]

use rym_errors::{Diagnostic, Level};
use rym_span::Span;

use super::{token::LexerError, Token};

#[track_caller]
fn assert_results(src: &str, expect: &[(Token, Span)], errors: &[LexerError]) {
	// TODO Make tests work or add insta snapshot testing
	// let got: Vec<_> = super::Lexer::new(src).collect();
	// println!("{got:#?}");
	// assert_eq!(expect, got);
	// assert_eq!(errors, &handler.collect());
}

#[track_caller]
fn assert_tokens(src: &str, expect: &[(Token, Span)]) {
	// let handler = DiagnosticHandler::default();
	// let got: Vec<Token> = super::Lexer::new(src, &handler).collect();
	// let errors = handler.collect();
	// assert_eq!(errors, vec![], "Expected no errors got: {errors:?}");
	// println!("{got:#?}");
	// assert_eq!(got, expect);
}

#[track_caller]
fn assert_diagnostics(src: &str, expect: &[Diagnostic]) {
	// let handler = DiagnosticHandler::default();
	// let _: Vec<_> = super::Lexer::new(src, &handler).collect();
	// let errors = handler.collect();
	// println!("{errors:#?}",);
	// assert_eq!(errors, expect)
}

#[track_caller]
fn assert_token_kinds(src: &str, expect: &[Token]) {
	// let handler = DiagnosticHandler::default();
	// let got: Vec<TokenKind> = super::Lexer::new(src, &handler).map(|token| token.kind).collect();
	// let errors = handler.collect();
	// assert_eq!(errors, vec![], "Expected no errors got: {errors:?}");
	// println!("{got:#?}");
	// assert_eq!(got, expect);
}

#[test]
fn empty() {
	assert_results("", &[], &[])
}

#[test]
fn expr_if() {
	let src = r#"if test == true { "is correct" } else { "is not correct" };"#;
	assert_tokens(
		src,
		&[
			(Token::If, Span::new(0, 2)),
			(Token::Ident("test".to_string()), Span::new(3, 7)),
			(Token::EqEq, Span::new(8, 10)),
			(Token::Ident("true".to_string()), Span::new(11, 15)),
			(Token::OpenBrace, Span::new(16, 17)),
			(Token::String("is correct".into()), Span::new(18, 30)),
			(Token::CloseBrace, Span::new(31, 32)),
			(Token::Else, Span::new(33, 37)),
			(Token::OpenBrace, Span::new(38, 39)),
			(Token::String("is not correct".into()), Span::new(40, 56)),
			(Token::CloseBrace, Span::new(57, 58)),
			(Token::Semi, Span::new(58, 59)),
		],
	);
	assert_token_kinds(
		src,
		&[
			Token::If,
			Token::Ident("test".to_string()),
			Token::EqEq,
			Token::Ident("true".to_string()),
			Token::OpenBrace,
			Token::String("is correct".into()),
			Token::CloseBrace,
			Token::Else,
			Token::OpenBrace,
			Token::String("is not correct".into()),
			Token::CloseBrace,
			Token::Semi,
		],
	);
}

#[test]
fn literals() {
	assert_token_kinds(
		r#"999_999_999 1.284_324_68 'c' '\n' '\x41' '\u24B6' '\u8DEF' "Hello World!\n""#,
		&[
			Token::Int(999_999_999),
			Token::Float("1.284_324_68".into()),
			Token::Char('c'),
			Token::Char('\n'),
			Token::Char('A'),
			Token::Char('Ⓐ'),
			Token::Char('路'),
			Token::String("Hello World!\n".into()),
		],
	)
}

#[test]
fn path() {
	assert_token_kinds(
		r#"std::__test::me: ::global_type"#,
		&[
			Token::Ident("std".to_string()),
			Token::ColonColon,
			Token::Ident("__test".to_string()),
			Token::ColonColon,
			Token::Ident("me".to_string()),
			Token::Colon,
			Token::ColonColon,
			Token::Ident("global_type".to_string()),
		],
	)
}

#[test]
fn operators() {
	assert_token_kinds(
		r#"| || & && + += - -= * *= / /= % %= = == ! != < <= > >="#,
		&[
			Token::Pipe,
			Token::PipePipe,
			Token::And,
			Token::AndAnd,
			Token::Plus,
			Token::PlusEq,
			Token::Minus,
			Token::MinusEq,
			Token::Star,
			Token::StarEq,
			Token::Slash,
			Token::SlashEq,
			Token::Percent,
			Token::PercentEq,
			Token::Eq,
			Token::EqEq,
			Token::Bang,
			Token::BangEq,
			Token::LessThan,
			Token::LessThanEq,
			Token::GreaterThan,
			Token::GreaterThanEq,
		],
	)
}

#[test]
fn reserved_char() {
	assert_diagnostics(
		"@^$#~?",
		&[
			Diagnostic::new_spanned(Level::Error, "Reserved character", Span::new(0, 1)),
			Diagnostic::new_spanned(Level::Error, "Reserved character", Span::new(1, 2)),
			Diagnostic::new_spanned(Level::Error, "Reserved character", Span::new(2, 3)),
			Diagnostic::new_spanned(Level::Error, "Reserved character", Span::new(3, 4)),
			Diagnostic::new_spanned(Level::Error, "Reserved character", Span::new(4, 5)),
			Diagnostic::new_spanned(Level::Error, "Reserved character", Span::new(5, 6)),
		],
	)
}

#[test]
fn invalid_char() {
	assert_diagnostics(
		"²³€",
		&[
			Diagnostic::new_spanned(Level::Error, "Invalid character", Span::new(0, 2)),
			Diagnostic::new_spanned(Level::Error, "Invalid character", Span::new(2, 4)),
			Diagnostic::new_spanned(Level::Error, "Invalid character", Span::new(4, 7)),
		],
	)
}

#[test]
fn unterminated() {
	assert_diagnostics(
		"/* *",
		&[
			Diagnostic::new_spanned(Level::Error, "Unterminated block comment", Span::new(0, 4))
				.sub_diagnostic(
					Level::Note,
					None,
					"Missing trailing `*/` to terminate the block comment",
				),
		],
	);
	assert_diagnostics(
		"\"Hello World\n",
		&[Diagnostic::new_spanned(
			Level::Error,
			"Unterminated string literal",
			Span::new(0, 13),
		)
		.sub_diagnostic(
			Level::Note,
			None,
			"Missing trailing `\"` to terminate the string literal",
		)],
	);
	assert_diagnostics(
		"'\n'",
		&[
			Diagnostic::new_spanned(
				Level::Error,
				"Unterminated character literal",
				Span::new(0, 2),
			)
			.sub_diagnostic(
				Level::Note,
				None,
				"Missing trailing `'` to terminate the character literal",
			),
			Diagnostic::new_spanned(
				Level::Error,
				"Unterminated character literal",
				Span::new(2, 3),
			)
			.sub_diagnostic(
				Level::Note,
				None,
				"Missing trailing `'` to terminate the character literal",
			),
		],
	);
}

#[test]
fn call() {
	assert_tokens(
		"empty_call()",
		&[
			(Token::Ident("empty_call".to_string()), Span::new(0, 10)),
			(Token::OpenParen, Span::new(10, 11)),
			(Token::CloseParen, Span::new(11, 12)),
		],
	);
	assert_tokens(
		"call(arg_1: float, arg_2: bool = true)",
		&[
			(Token::Ident("call".to_string()), Span::new(0, 4)),
			(Token::OpenParen, Span::new(4, 5)),
			(Token::Ident("arg_1".to_string()), Span::new(5, 10)),
			(Token::Colon, Span::new(10, 11)),
			(Token::Ident("float".to_string()), Span::new(12, 17)),
			(Token::Comma, Span::new(17, 18)),
			(Token::Ident("arg_2".to_string()), Span::new(19, 24)),
			(Token::Colon, Span::new(24, 25)),
			(Token::Ident("bool".to_string()), Span::new(26, 30)),
			(Token::Eq, Span::new(31, 32)),
			(Token::Ident("true".to_string()), Span::new(33, 37)),
			(Token::CloseParen, Span::new(37, 38)),
		],
	);
	assert_token_kinds(
		"call(arg_1: float, arg_2: bool = true)",
		&[
			Token::Ident("call".to_string()),
			Token::OpenParen,
			Token::Ident("arg_1".to_string()),
			Token::Colon,
			Token::Ident("float".to_string()),
			Token::Comma,
			Token::Ident("arg_2".to_string()),
			Token::Colon,
			Token::Ident("bool".to_string()),
			Token::Eq,
			Token::Ident("true".to_string()),
			Token::CloseParen,
		],
	)
}

#[test]
fn function() {
	assert_token_kinds(
		"fn add(a: int, b: int) -> int { a + b }",
		&[
			Token::Func,
			Token::Ident("add".to_string()),
			Token::OpenParen,
			Token::Ident("a".to_string()),
			Token::Colon,
			Token::Ident("int".to_string()),
			Token::Comma,
			Token::Ident("b".to_string()),
			Token::Colon,
			Token::Ident("int".to_string()),
			Token::CloseParen,
			Token::ThinArrow,
			Token::Ident("int".to_string()),
			Token::OpenBrace,
			Token::Ident("a".to_string()),
			Token::Plus,
			Token::Ident("b".to_string()),
			Token::CloseBrace,
		],
	)
}

#[test]
fn unclosed() {
	// TODO Add multiline variants of tests
	assert_results(
		"({[",
		&[
			(Token::OpenParen, Span::new(0, 1)),
			(Token::OpenBrace, Span::new(1, 2)),
			(Token::OpenBracket, Span::new(2, 3)),
		],
		&[],
	);
	assert_results(
		"{ a + b",
		&[
			(Token::OpenBrace, Span::new(0, 1)),
			(Token::Ident("a".to_string()), Span::new(2, 3)),
			(Token::Plus, Span::new(4, 5)),
			(Token::Ident("b".to_string()), Span::new(6, 7)),
		],
		&[],
	);
	assert_results(
		"{ a + (b)",
		&[
			(Token::OpenBrace, Span::new(0, 1)),
			(Token::Ident("a".to_string()), Span::new(2, 3)),
			(Token::Plus, Span::new(4, 5)),
			(Token::OpenParen, Span::new(6, 7)),
			(Token::Ident("b".to_string()), Span::new(7, 8)),
			(Token::CloseParen, Span::new(8, 9)),
		],
		&[],
	)
}
