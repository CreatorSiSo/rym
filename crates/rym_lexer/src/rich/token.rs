use rym_span::Span;

#[derive(Debug, PartialEq)]
pub struct LexerError {
	span: Span,
	msg: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
#[non_exhaustive]
pub enum Token {
	// Punctuation
	/// `;`
	Semi,
	/// `:`
	Colon,
	/// `::`
	ColonColon,
	/// `.`
	Dot,
	/// `..`
	DotDot,
	/// `,`
	Comma,

	// Operator like
	/// `|`
	Pipe,
	/// `||`
	PipePipe,
	/// `&`
	And,
	/// `&&`
	AndAnd,
	/// `+`
	Plus,
	/// `+=`
	PlusEq,
	/// `-`
	Minus,
	/// `-=`
	MinusEq,
	/// `*`
	Star,
	/// `*=`
	StarEq,
	/// `/`
	Slash,
	/// `/=`
	SlashEq,
	/// `%`
	Percent,
	/// `%=`
	PercentEq,
	/// `!`
	Bang,
	/// `!=`
	BangEq,
	/// `=`
	Eq,
	/// `==`
	EqEq,
	/// `<`
	LessThan,
	/// `<=`
	LessThanEq,
	/// `>`
	GreaterThan,
	/// `>=`
	GreaterThanEq,
	/// `->`
	ThinArrow,
	/// `=>`
	FatArrow,

	// Keywords
	/// `const`
	Const,
	/// `else`
	Else,
	/// `enum`
	Enum,
	/// `for`
	For,
	/// `func`
	Func,
	/// `if`
	If,
	/// `impl`
	Impl,
	/// `in`
	In,
	/// `loop`
	Loop,
	/// `mod`
	Mod,
	/// `mut`
	Mut,
	/// `struct`
	Struct,
	/// `then`
	Then,
	/// `trait`
	Trait,
	/// `type`
	Type,
	/// `use`
	Use,
	/// `while`
	While,

	/// Delimiter token.
	/// `(`
	OpenParen,
	/// `)`
	CloseParen,
	/// `{`
	OpenBrace,
	/// `}`
	CloseBrace,
	/// `[`
	OpenBracket,
	/// `]`
	CloseBracket,

	/// Indentifier token: `some_thing`, `test`
	Ident(String),

	/// `36_254`
	Int(u64),
	/// `0.2346`
	Float(u64, u64),
	/// `'a'`, `'\n'`
	Char(char),
	/// `"abcde"`, `"Hello World!\n"`
	String(String),

	/// Well thats where it ends...
	Eof,
}

impl std::fmt::Display for Token {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		let mut tmp = String::new();
		fn hoist_up<'a>(tmp: &'a mut String, data: String) -> &'a str {
			*tmp = data;
			&*tmp
		}

		let str = match self {
			Token::Semi => ";",
			Token::Colon => ":",
			Token::ColonColon => "::",
			Token::Dot => ".",
			Token::DotDot => "..",
			Token::Comma => ",",
			Token::Pipe => "|",
			Token::PipePipe => "||",
			Token::And => "&",
			Token::AndAnd => "&&",
			Token::Plus => "+",
			Token::PlusEq => "+=",
			Token::Minus => "-",
			Token::MinusEq => "-=",
			Token::Star => "*",
			Token::StarEq => "*=",
			Token::Slash => "/",
			Token::SlashEq => "/=",
			Token::Percent => "%",
			Token::PercentEq => "%=",
			Token::Bang => "!",
			Token::BangEq => "!=",
			Token::Eq => "=",
			Token::EqEq => "==",
			Token::LessThan => "<",
			Token::LessThanEq => "<=",
			Token::GreaterThan => ">",
			Token::GreaterThanEq => ">=",
			Token::ThinArrow => "->",
			Token::FatArrow => "=>",

			Token::Const => "const",
			Token::Else => "else",
			Token::Enum => "enum",
			Token::For => "for",
			Token::Func => "func",
			Token::If => "if",
			Token::Impl => "impl",
			Token::In => "in",
			Token::Loop => "loop",
			Token::Mod => "mod",
			Token::Mut => "mut",
			Token::Struct => "struct",
			Token::Then => "then",
			Token::Trait => "trait",
			Token::Type => "type",
			Token::Use => "use",
			Token::While => "while",

			Token::OpenParen => "(",
			Token::CloseParen => ")",

			Token::OpenBrace => "{",
			Token::CloseBrace => "",

			Token::OpenBracket => "[",
			Token::CloseBracket => "]",

			Token::Ident(name) => name,
			Token::Int(inner) => hoist_up(&mut tmp, inner.to_string()),
			Token::Float(lhs, rhs) => hoist_up(&mut tmp, format!("{lhs}.{rhs}")),
			Token::Char(inner) => hoist_up(&mut tmp, inner.to_string()),
			Token::String(inner) => inner,
			Token::Eof => "<eof>",
		};
		f.write_str(str)
	}
}

const KEYWORDS_LEN: usize = 17;
pub const KEYWORDS_MAP: ([&str; KEYWORDS_LEN], [Token; KEYWORDS_LEN]) = (
	[
		// a
		// b
		"const", // c
		// d
		"else", // e
		"enum", // e
		"for", "func", // f
		// g
		// h
		"if", "impl", "in", // i
		// j
		// k
		"loop", // l
		"mod", "mut", // m
		// n
		// o
		// p
		// q
		// r
		"struct", // s
		"then", "trait", "type", // t
		"use",  // u
		// v
		"while", // w
		         // x
		         // y
		         // z
	],
	[
		Token::Const,
		Token::Else,
		Token::Enum,
		Token::For,
		Token::Func,
		Token::If,
		Token::Impl,
		Token::In,
		Token::Loop,
		Token::Mod,
		Token::Mut,
		Token::Struct,
		Token::Then,
		Token::Trait,
		Token::Type,
		Token::Use,
		Token::While,
	],
);
