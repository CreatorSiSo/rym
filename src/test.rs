#![cfg(test)]

use super::*;
use chumsky::Stream;
use indoc::indoc;
use rym_lexer::rich::Lexer;

fn parse_expr(src: &'static str) -> (Option<(Expr, Span)>, Vec<Report>) {
	let token_stream = Stream::from_iter(0..0, Lexer::new(src));
	let (ast, mut errors) = expr_parser().parse_recovery(token_stream);

	errors.sort_by(|l, r| l.span.start.cmp(&r.span.start));
	let reports = errors.into_iter().map(|err| Report::from(err)).collect();

	(ast, reports)
}

macro_rules! insta_assert_parser {
	($parser:expr; $($src:expr),+ $(,)?) => {
		$({
			let (ast, reports) = parse_expr($src);
			let snapshot = format!(
				"--- Input ---\n{}\n---\n\n{:#?}\n\n--- Errors ---\n{:#?}\n---",
				$src, &ast, &reports
			);
			insta::assert_snapshot!(&snapshot);
		})*
	};
}

#[test]
fn literal_expressions() {
	insta_assert_parser! {
		expr_parser();
		"823_472_340",
		"823_472_340_983_273_327",
		"1.0",
		"0.",
		"444.444",
		"9_999_999_999_999.999_999_999",
		"'a'",
		"'\\n'",
		"'\\t'",
		"'\\r'",
		r#" "abc" "#,
		r#" "Hello World!\n" "#,
	}
}

#[test]
fn record_expressions() {
	insta_assert_parser! {
		expr_parser();
		r#"Record { name: "Record", fields: todo() }"#,
		r#".{ name: "Record", fields: todo() }"#,
		r#".{ }"#,
	}
}

#[test]
fn simple_expressions() {
	insta_assert_parser! {
		expr_parser();
		"indentifier_1",
		"(grouped_ident)",
		"func_name(2, \"Do stuff!!\", true)",
		"(make_func())()()()",
		"(((wrapped)))()",
		"(if true then 1 else 2) + 3",
	}
}

#[test]
fn unary_expressions() {
	insta_assert_parser! {
		expr_parser();
		"-255",
		"-1.0",
		"-0.",
		"---3",

		"!true",
		"!false",
		"!!!true",
		"!(lhs == rhs)",
		"!lhs == rhs",
	}
}

#[test]
fn binary_expressions() {
	insta_assert_parser! {
		expr_parser();
		"lhs == rhs",
		"lhs != rhs",
		"lhs + rhs",
		"lhs - rhs",
		"lhs * rhs",
		"lhs / rhs",


		"1 + 2 - 3 * 4 / 5",
		"1 == 1 + 2 / 3 != 0",
	}
}

#[test]
fn block_expressions() {
	insta_assert_parser! {
		expr_parser();
		indoc!("
		{
			testing;
			testing;
		}"),
		"{ call_me(); and_me(); }",
		"{ }",
	}
}

#[test]
fn variables() {
	insta_assert_parser! {
		expr_parser();
		indoc!(r#"
		{
			const msg = "Hello World!\n";
			mut counter = 0;

			// TODO: count from 0 to 10 and print f"{counter}: Hello World!\n" every time
		}"#),
	}
}

#[test]
fn if_expressions() {
	insta_assert_parser! {
		expr_parser();
		r#"if true then { print("Hello Universe!"); } else { print("Hello World!"); }"#,
		r#"if (true == false) then ("Hello Universe!") else ("Hello World!")"#,
		r#"if true then "Hello Universe!" else "Hello World!""#,
		indoc!(r#"
			if true then "Hello Universe!"
			else "Hello World!""#),
		indoc!(r#"
			if stuff_goes_right(1, 2, 9) then {
				print("Hello Universe!");
				20;
			} else {
				print("Hello World!");
				10;
			}"#),
	}
}

#[test]
fn recover_group() {
	insta_assert_parser!(
		expr_parser();

		// invalid semicolon
		"(testing;)",

		// missing rhs
		"(1 +)",
		"( / 888)",

		// unclosed group
		"((testing)",
		"(testing}",
	);
}

#[test]
fn recover_block() {
	insta_assert_parser!(
		expr_parser();

		// missing semicolon
		"{ testing }",

		// unclosed
		"{ (testing; }",
		// TODO: improve error recovery so that it keeps at least one block
		"{ {testing; }",

		// additional closing delimiter
		"{ testing ]}",
		"{ testing )}",
		"{ testing }}",

		// wrong closing delimiter
		"{ testing ]",
		"{ testing )",
	);
}

#[test]
fn recover_call() {
	insta_assert_parser!(
		expr_parser();

		// missing args
		"testing(,)",
		"testing(,)(correct_one, 2)(, aerr, \"stuff\")",

		// missing commas
		"testing(1 2 3)",
		"testing(1 + / 2)",

		// unclosed
		"testing(1, {2, 3)",
		// "testing(1, (2, 3)",
	);
}
