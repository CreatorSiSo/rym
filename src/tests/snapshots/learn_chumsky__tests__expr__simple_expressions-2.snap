---
source: src/test.rs
expression: "&snapshot"
---
--- Input ---
(grouped_ident)
---

Some(
    (
        Ident("grouped_ident"),
        1..14,
    ),
)

--- Errors ---
[]
---
